File: plugins/vibeassist/skills/vibeassist-map/scripts/routes_react_router.py
#!/usr/bin/env python3
"""Phase 1 route enumeration for react-router apps (Lovable-style) — Fix 2.

Routes are JSX in a tree, not files, so this PARSES rather than globs:
  - <Routes>/<Route> JSX trees (BrowserRouter style, Lovable's default)
  - createBrowserRouter([...]) object trees
  - relative paths COMPOSE down the nesting: <Route path="edit"> inside
    <Route path="projects/:id"> is /projects/:id/edit
  - <Route index> means the parent's own path, not a new segment
  - a <Route> with children and no element is a LAYOUT, not a route
    (the same layout-is-not-a-route guard Phase 1 has for file-based routers)
  - <Navigate to> => audience "redirect" with redirect_to
  - path="*" => the not-found page (user-facing, flagged)
  - components arrive via lazy(() => import('./pages/Foo')) or static imports;
    both are resolved (through tsconfig path aliases) to a source_file

Usage:
    python scripts/routes_react_router.py --repo-root . -o map/_routes.json
    (optionally: --entry src/App.tsx to skip the search)
"""
import argparse
import json
import pathlib
import re
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from harvest import load_aliases, resolve_import, route_slug  # noqa: E402

try:
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception:
    pass

ROUTE_OPEN = re.compile(r"<Route\b")
ROUTE_CLOSE = re.compile(r"</Route\s*>")


def tokenize_route_tags(text: str):
    """Yield ('open'|'selfclose'|'close', attr_text, pos). A JSX attr like
    element={<AppShell />} contains '>' inside braces, so the tag's true end
    must be found by tracking {} depth - a plain [^>]* regex truncates it."""
    i = 0
    while i < len(text):
        mo = ROUTE_OPEN.search(text, i)
        mc = ROUTE_CLOSE.search(text, i)
        if not mo and not mc:
            return
        if mc and (not mo or mc.start() < mo.start()):
            yield ("close", "", mc.start())
            i = mc.end()
            continue
        j, depth = mo.end(), 0
        while j < len(text):
            ch = text[j]
            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
            elif ch == ">" and depth == 0:
                break
            j += 1
        attr_text = text[mo.end():j]
        kind = "selfclose" if attr_text.rstrip().endswith("/") else "open"
        yield (kind, attr_text.rstrip().rstrip("/"), mo.start())
        i = j + 1
ATTR_PATH = re.compile(r"""\bpath\s*=\s*(?:["']([^"']+)["']|\{\s*["']([^"']+)["']\s*\})""")
ATTR_INDEX = re.compile(r"\bindex\b(?!\s*=\s*\{?\s*false)")
ATTR_ELEMENT = re.compile(r"\belement\s*=\s*\{\s*<\s*([A-Za-z_]\w*)")
ATTR_NAVIGATE = re.compile(r"""<\s*Navigate\b[^>]*\bto\s*=\s*(?:["']([^"']+)["']|\{\s*["']([^"']+)["']\s*\})""")
LAZY_IMPORT = re.compile(r"""(?:const|let|var)\s+(\w+)\s*=\s*(?:React\.)?lazy\(\s*\(\)\s*=>\s*import\(\s*['"]([^'"]+)['"]\s*\)""")
STATIC_IMPORT = re.compile(r"""import\s+(?:(\w+)|\{([^}]+)\})\s+from\s+['"]([^'"]+)['"]""")


def norm_param(path: str) -> str:
    return path  # react-router already uses :param — the canonical form


def join_paths(parent: str, child: str) -> str:
    if child.startswith("/"):
        return child
    if not parent:
        return "/" + child
    return (parent.rstrip("/") + "/" + child) if child else parent


def build_import_map(text: str):
    """component name -> import spec (lazy and static)."""
    imp = {}
    for name, spec in LAZY_IMPORT.findall(text):
        imp[name] = spec
    for default, named, spec in STATIC_IMPORT.findall(text):
        if default:
            imp[default] = spec
        for n in named.split(","):
            n = n.strip().split(" as ")[-1].strip()
            if n:
                imp[n] = spec
    return imp


def attrs_of(attr_text: str):
    m = ATTR_PATH.search(attr_text)
    path = (m.group(1) or m.group(2)) if m else None
    el = ATTR_ELEMENT.search(attr_text)
    nav = ATTR_NAVIGATE.search(attr_text)
    return {
        "path": path,
        "index": bool(ATTR_INDEX.search(re.sub(ATTR_PATH.pattern, "", attr_text))),
        "component": el.group(1) if el else None,
        "navigate_to": (nav.group(1) or nav.group(2)) if nav else None,
    }


def parse_jsx(text: str):
    """Walk <Route> tags with a stack; emit route dicts for leaves and
    element-bearing nodes; layouts (children, no element) are chain-only."""
    routes = []
    stack = []  # each: {"attrs":…, "children_seen":bool}

    def emit(node, ancestors):
        a = node["attrs"]
        full = ""
        for anc in ancestors:
            p = anc["attrs"]["path"]
            if p:
                full = join_paths(full, p)
        if a["index"]:
            path = full or "/"
        elif a["path"] is not None:
            path = join_paths(full, a["path"])
        else:
            return  # pathless, non-index, no path: pure layout wrapper
        entry = {"path": path or "/",
                 "component": a["component"],
                 "layout_chain": [anc["attrs"]["component"] for anc in ancestors if anc["attrs"]["component"]],
                 "audience": "user"}
        if a["navigate_to"]:
            entry["audience"] = "redirect"
            entry["redirect_to"] = join_paths(full, a["navigate_to"]) if not a["navigate_to"].startswith("/") else a["navigate_to"]
        if a["path"] == "*" or path.endswith("/*") and a["path"] == "*":
            entry["not_found"] = True
        routes.append(entry)

    for kind, attrs, _pos in tokenize_route_tags(text):
        if kind == "close":
            if stack:
                node = stack.pop()
                # closing a Route that had children: it's a layout if no element,
                # or a layout-with-UI if element+children — either way children cover it
                if not node["children_seen"] and (node["attrs"]["path"] is not None or node["attrs"]["index"] or node["attrs"]["navigate_to"]):
                    emit(node, stack)
            continue
        node = {"attrs": attrs_of(attrs), "children_seen": False}
        if stack:
            stack[-1]["children_seen"] = True
        if kind == "selfclose":
            emit(node, stack)
        else:
            stack.append(node)
    return routes


def split_objects(arr_text: str):
    """Split a JS array literal's top-level {...} objects."""
    objs, depth, start = [], 0, None
    for i, ch in enumerate(arr_text):
        if ch == "{":
            if depth == 0:
                start = i
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0 and start is not None:
                objs.append(arr_text[start + 1:i])
                start = None
    return objs


def find_children_span(obj_text: str):
    m = re.search(r"\bchildren\s*:\s*\[", obj_text)
    if not m:
        return None, obj_text
    depth, i = 1, m.end()
    while i < len(obj_text) and depth:
        if obj_text[i] == "[":
            depth += 1
        elif obj_text[i] == "]":
            depth -= 1
        i += 1
    return obj_text[m.end():i - 1], obj_text[:m.start()] + obj_text[i:]


def parse_router_objects(arr_text: str, parent_path: str, chain, routes):
    for obj in split_objects(arr_text):
        children_text, own = find_children_span(obj)
        a = attrs_of(own.replace("path:", "path=").replace("element:", "element={").replace("index:", "index "))
        # object syntax needs its own field regexes; redo path/index simply:
        pm = re.search(r"""\bpath\s*:\s*["']([^"']+)["']""", own)
        a["path"] = pm.group(1) if pm else None
        a["index"] = bool(re.search(r"\bindex\s*:\s*true", own))
        em = re.search(r"\belement\s*:\s*<\s*([A-Za-z_]\w*)", own)
        a["component"] = em.group(1) if em else None
        nm = ATTR_NAVIGATE.search(own)
        a["navigate_to"] = (nm.group(1) or nm.group(2)) if nm else None
        if a["index"]:
            full = parent_path or "/"
        elif a["path"] is not None:
            full = join_paths(parent_path, a["path"])
        else:
            full = parent_path  # pathless layout
        if children_text is not None:
            new_chain = chain + ([a["component"]] if a["component"] else [])
            parse_router_objects(children_text, full if a["path"] is not None else parent_path, new_chain, routes)
            continue  # layout: not itself a route
        if a["path"] is None and not a["index"]:
            continue
        entry = {"path": full or "/", "component": a["component"], "layout_chain": chain, "audience": "user"}
        if a["navigate_to"]:
            entry["audience"] = "redirect"
            entry["redirect_to"] = a["navigate_to"]
        if a["path"] == "*":
            entry["not_found"] = True
        routes.append(entry)


def find_entry(repo: pathlib.Path):
    cands = []
    for name in ("src/App.tsx", "src/App.jsx", "src/main.tsx", "src/main.jsx", "src/router.tsx", "src/routes.tsx", "app/App.tsx"):
        f = repo / name
        if f.is_file():
            cands.append(f)
    for f in repo.glob("src/**/*.[tj]sx"):
        if "node_modules" not in f.parts and f not in cands:
            try:
                t = f.read_text(encoding="utf-8", errors="replace")
            except OSError:
                continue
            if "createBrowserRouter" in t or "<Routes" in t:
                cands.append(f)
    for f in cands:
        t = f.read_text(encoding="utf-8", errors="replace")
        if "createBrowserRouter" in t or "<Routes" in t or "<Route" in t:
            return f, t
    return None, None


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--repo-root", default=".")
    ap.add_argument("--entry", default=None, help="router file, e.g. src/App.tsx")
    ap.add_argument("-o", "--out", default="map/_routes.json")
    args = ap.parse_args()
    repo = pathlib.Path(args.repo_root).resolve()
    aliases = load_aliases(repo)

    if args.entry:
        entry = repo / args.entry
        text = entry.read_text(encoding="utf-8", errors="replace")
    else:
        entry, text = find_entry(repo)
        if entry is None:
            sys.exit("No react-router registration found (createBrowserRouter or <Routes>). Wrong repo or wrong strategy.")

    routes = []
    m = re.search(r"createBrowserRouter\s*\(\s*\[", text)
    if m:
        depth, i = 1, m.end()
        while i < len(text) and depth:
            if text[i] == "[":
                depth += 1
            elif text[i] == "]":
                depth -= 1
            i += 1
        parse_router_objects(text[m.end():i - 1], "", [], routes)
    routes.extend(parse_jsx(text))

    imp = build_import_map(text)
    out_routes, seen = [], set()
    for r in routes:
        if r["path"] in seen:
            continue
        seen.add(r["path"])
        comp = r.pop("component", None)
        src_file = None
        if comp and comp in imp:
            resolved = resolve_import(entry, imp[comp], repo, aliases)
            if resolved:
                src_file = str(resolved.relative_to(repo)).replace("\\", "/")
        out_routes.append({"path": r["path"],
                           "source_file": src_file or str(entry.relative_to(repo)).replace("\\", "/"),
                           "component": comp,
                           "layout_chain": r.get("layout_chain", []),
                           "auth_required": None,  # react-router guards are wrapper components; decide per-app in Phase 1 prose
                           "audience": r.get("audience", "user"),
                           **({"redirect_to": r["redirect_to"]} if "redirect_to" in r else {}),
                           **({"not_found": True} if r.get("not_found") else {}),
                           "slug": route_slug(r["path"])})
    outp = pathlib.Path(args.out)
    outp.parent.mkdir(parents=True, exist_ok=True)
    outp.write_text(json.dumps(out_routes, indent=2), encoding="utf-8")
    nred = sum(1 for r in out_routes if r["audience"] == "redirect")
    print(f"react-router: {len(out_routes)} routes -> {outp} ({nred} redirects) from {entry.name}")
    if not out_routes:
        sys.exit("Router file found but zero routes parsed - inspect it by hand before continuing.")


if __name__ == "__main__":
    main()

File: plugins/vibeassist/skills/vibeassist-map/scripts/routes_react_router.mjs
#!/usr/bin/env node
// Phase 1 route enumeration for react-router apps (Lovable-style) - Fix 2.
//
// Faithful ESM port of routes_react_router.py. Behaviour (CLI args, input
// reading, output file, stdout text, exit codes) is byte-for-byte identical to
// the Python. The three shared helpers load_aliases / resolve_import /
// route_slug are imported from harvest.mjs exactly as the Python module imports
// them from harvest.py - they are NOT reimplemented here.

import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { dumps } from './_pyjson.mjs'
import { load_aliases, resolve_import, route_slug } from './harvest.mjs'

// ---------------------------------------------------------------------------
// Python-ism helpers
// ---------------------------------------------------------------------------

// CPython str whitespace set (Py_UNICODE_ISSPACE), used by str.rstrip()/strip().
const PY_WS = '\\t\\n\\x0b\\x0c\\r\\x1c\\x1d\\x1e\\x1f \\x85\\xa0\\u1680\\u2000-\\u200a\\u2028\\u2029\\u202f\\u205f\\u3000'
const STRIP_R = new RegExp('[' + PY_WS + ']+$')

function pyRstrip(s) {
  return s.replace(STRIP_R, '')
}
// str.rstrip(chars): strip trailing chars only.
function rstripChars(s, chars) {
  const set = new Set(chars)
  let j = s.length
  while (j > 0 && set.has(s[j - 1])) j--
  return s.slice(0, j)
}

// Pattern.search(string, pos): global regex with lastIndex set. `\b`/lookbehind
// still see the full string (like Python's pos parameter, which is not slicing).
function searchFrom(reGlobal, text, pos) {
  reGlobal.lastIndex = pos
  const m = reGlobal.exec(text)
  return m
}

// ---------------------------------------------------------------------------
// Filesystem helpers (pathlib semantics on string paths)
// ---------------------------------------------------------------------------
function readTextReplace(p) {
  // pathlib read_text(encoding='utf-8', errors='replace')
  return fs.readFileSync(p).toString('utf8')
}
function isFile(p) {
  try { return fs.statSync(p).isFile() } catch { return false }
}
function isDirPath(p) {
  try { return fs.statSync(p).isDirectory() } catch { return false }
}
function pathExists(p) {
  try { fs.statSync(p); return true } catch { return false }
}
function relTo(repo, p) {
  return path.relative(repo, p).split(path.sep).join('/')
}
function baseName(p) {
  return path.basename(p)
}
// f.parts membership test for 'node_modules' on an absolute posix path.
function partsHasNodeModules(absPath) {
  // Split on the OS separator too: glob() builds these with path.join, so on
  // Windows they carry backslashes and a '/'-only split never sees the segment.
  return absPath.split(path.sep).join('/').split('/').includes('node_modules')
}

// --- pathlib-compatible glob (repo.glob("src/**/*.[tj]sx")) -----------------
function scandir(dir) {
  try { return fs.readdirSync(dir, { withFileTypes: true }) } catch { return [] }
}
function direntIsDir(dir, ent) {
  if (ent.isSymbolicLink()) {
    try { return fs.statSync(path.join(dir, ent.name)).isDirectory() } catch { return false }
  }
  return ent.isDirectory()
}
function* iterateDirectories(parent) {
  yield parent
  for (const ent of scandir(parent)) {
    if (ent.isDirectory() && !ent.isSymbolicLink()) {
      yield* iterateDirectories(path.join(parent, ent.name))
    }
  }
}
function isWildcard(pat) {
  return pat.includes('*') || pat.includes('?') || pat.includes('[')
}
// fnmatch.translate for a single path segment: handles *, ?, and [..] classes
// (Python char classes are NOT escaped away). Anchored + dotall like pathlib.
function fnmatchTranslate(pat) {
  let res = ''
  let i = 0
  const n = pat.length
  while (i < n) {
    const c = pat[i]
    i += 1
    if (c === '*') {
      res += '.*'
    } else if (c === '?') {
      res += '.'
    } else if (c === '[') {
      let j = i
      if (j < n && pat[j] === '!') j += 1
      if (j < n && pat[j] === ']') j += 1
      while (j < n && pat[j] !== ']') j += 1
      if (j >= n) {
        res += '\\['
      } else {
        let stuff = pat.slice(i, j).replace(/\\/g, '\\\\')
        i = j + 1
        if (stuff[0] === '!') stuff = '^' + stuff.slice(1)
        else if (stuff[0] === '^' || stuff[0] === '[') stuff = '\\' + stuff
        res += '[' + stuff + ']'
      }
    } else if ('.^$+{}()|\\'.includes(c)) {
      res += '\\' + c
    } else {
      res += c
    }
  }
  return new RegExp('^(?:' + res + ')$', 's')
}
function makeSelector(parts) {
  if (parts.length === 0) {
    return function* (p) { yield p }
  }
  const pat = parts[0]
  const rest = parts.slice(1)
  const successor = makeSelector(rest)
  const dironly = rest.length > 0
  if (pat === '**') {
    return function* (parent) {
      const yielded = new Set()
      for (const start of iterateDirectories(parent)) {
        for (const p of successor(start)) {
          if (!yielded.has(p)) { yielded.add(p); yield p }
        }
      }
    }
  } else if (isWildcard(pat)) {
    const re = fnmatchTranslate(pat)
    return function* (parent) {
      for (const ent of scandir(parent)) {
        if (dironly && !direntIsDir(parent, ent)) continue
        if (re.test(ent.name)) yield* successor(path.join(parent, ent.name))
      }
    }
  } else {
    return function* (parent) {
      const child = path.join(parent, pat)
      const ok = dironly ? isDirPath(child) : pathExists(child)
      if (ok) yield* successor(child)
    }
  }
}
function glob(base, pattern) {
  return [...makeSelector(pattern.split('/'))(base)]
}

// PurePosixPath(str) normalisation for the printed / written output path.
function pyPosixStr(p) {
  const isAbs = p.startsWith('/')
  const segs = p.split('/').filter((s) => s !== '' && s !== '.')
  let s = segs.join('/')
  if (isAbs) s = '/' + s
  if (s === '') s = isAbs ? '/' : '.'
  return s
}

// ===========================================================================
// routes_react_router.py port
// ===========================================================================

const ROUTE_OPEN = /<Route\b/g
const ROUTE_CLOSE = /<\/Route\s*>/g

function* tokenize_route_tags(text) {
  // Yield ['open'|'selfclose'|'close', attr_text, pos]. A JSX attr like
  // element={<AppShell />} contains '>' inside braces, so the tag's true end
  // must be found by tracking {} depth - a plain [^>]* regex truncates it.
  let i = 0
  while (i < text.length) {
    const mo = searchFrom(ROUTE_OPEN, text, i)
    const mc = searchFrom(ROUTE_CLOSE, text, i)
    if (!mo && !mc) return
    if (mc && (!mo || mc.index < mo.index)) {
      yield ['close', '', mc.index]
      i = mc.index + mc[0].length
      continue
    }
    let j = mo.index + mo[0].length
    let depth = 0
    while (j < text.length) {
      const ch = text[j]
      if (ch === '{') depth += 1
      else if (ch === '}') depth -= 1
      else if (ch === '>' && depth === 0) break
      j += 1
    }
    const attr_text = text.slice(mo.index + mo[0].length, j)
    const kind = pyRstrip(attr_text).endsWith('/') ? 'selfclose' : 'open'
    yield [kind, rstripChars(pyRstrip(attr_text), '/'), mo.index]
    i = j + 1
  }
}

const ATTR_PATH_SRC = '\\bpath\\s*=\\s*(?:["\']([^"\']+)["\']|\\{\\s*["\']([^"\']+)["\']\\s*\\})'
const ATTR_PATH = new RegExp(ATTR_PATH_SRC)
const ATTR_PATH_G = new RegExp(ATTR_PATH_SRC, 'g')
const ATTR_INDEX = /\bindex\b(?!\s*=\s*\{?\s*false)/
const ATTR_ELEMENT = /\belement\s*=\s*\{\s*<\s*([A-Za-z_]\w*)/
const ATTR_NAVIGATE = /<\s*Navigate\b[^>]*\bto\s*=\s*(?:["']([^"']+)["']|\{\s*["']([^"']+)["']\s*\})/
const LAZY_IMPORT = /(?:const|let|var)\s+(\w+)\s*=\s*(?:React\.)?lazy\(\s*\(\)\s*=>\s*import\(\s*['"]([^'"]+)['"]\s*\)/g
const STATIC_IMPORT = /import\s+(?:(\w+)|\{([^}]+)\})\s+from\s+['"]([^'"]+)['"]/g

function norm_param(pathStr) {
  return pathStr // react-router already uses :param - the canonical form
}

function join_paths(parent, child) {
  if (child.startsWith('/')) return child
  if (!parent) return '/' + child
  return child ? (rstripChars(parent, '/') + '/' + child) : parent
}

function build_import_map(text) {
  // component name -> import spec (lazy and static).
  const imp = {}
  for (const m of text.matchAll(LAZY_IMPORT)) {
    const name = m[1] ?? ''
    const spec = m[2] ?? ''
    imp[name] = spec
  }
  for (const m of text.matchAll(STATIC_IMPORT)) {
    const def = m[1] ?? ''
    const named = m[2] ?? ''
    const spec = m[3] ?? ''
    if (def) imp[def] = spec
    for (let n of named.split(',')) {
      const parts = n.split(' as ')
      n = pyStripLocal(parts[parts.length - 1])
      if (n) imp[n] = spec
    }
  }
  return imp
}

// str.strip() (default whitespace) for import-name cleanup.
const STRIP_L = new RegExp('^[' + PY_WS + ']+')
function pyStripLocal(s) {
  return s.replace(STRIP_L, '').replace(STRIP_R, '')
}

function attrs_of(attr_text) {
  const m = ATTR_PATH.exec(attr_text)
  const pathv = m ? ((m[1] ?? m[2]) ?? null) : null
  const el = ATTR_ELEMENT.exec(attr_text)
  const nav = ATTR_NAVIGATE.exec(attr_text)
  return {
    path: pathv,
    index: Boolean(ATTR_INDEX.exec(attr_text.replace(ATTR_PATH_G, ''))),
    component: el ? el[1] : null,
    navigate_to: nav ? ((nav[1] ?? nav[2]) ?? null) : null,
  }
}

function parse_jsx(text) {
  // Walk <Route> tags with a stack; emit route dicts for leaves and
  // element-bearing nodes; layouts (children, no element) are chain-only.
  const routes = []
  const stack = [] // each: {attrs, children_seen}

  function emit(node, ancestors) {
    const a = node.attrs
    let full = ''
    for (const anc of ancestors) {
      const p = anc.attrs.path
      if (p) full = join_paths(full, p)
    }
    let pathStr
    if (a.index) {
      pathStr = full || '/'
    } else if (a.path !== null) {
      pathStr = join_paths(full, a.path)
    } else {
      return // pathless, non-index, no path: pure layout wrapper
    }
    const entry = {
      path: pathStr || '/',
      component: a.component,
      layout_chain: ancestors.filter((anc) => anc.attrs.component).map((anc) => anc.attrs.component),
      audience: 'user',
    }
    if (a.navigate_to) {
      entry.audience = 'redirect'
      entry.redirect_to = !a.navigate_to.startsWith('/') ? join_paths(full, a.navigate_to) : a.navigate_to
    }
    if (a.path === '*' || (pathStr.endsWith('/*') && a.path === '*')) {
      entry.not_found = true
    }
    routes.push(entry)
  }

  for (const [kind, attrs] of tokenize_route_tags(text)) {
    if (kind === 'close') {
      if (stack.length) {
        const node = stack.pop()
        if (!node.children_seen && (node.attrs.path !== null || node.attrs.index || node.attrs.navigate_to)) {
          emit(node, stack)
        }
      }
      continue
    }
    const node = { attrs: attrs_of(attrs), children_seen: false }
    if (stack.length) stack[stack.length - 1].children_seen = true
    if (kind === 'selfclose') emit(node, stack)
    else stack.push(node)
  }
  return routes
}

function split_objects(arr_text) {
  // Split a JS array literal's top-level {...} objects.
  const objs = []
  let depth = 0
  let start = null
  for (let i = 0; i < arr_text.length; i++) {
    const ch = arr_text[i]
    if (ch === '{') {
      if (depth === 0) start = i
      depth += 1
    } else if (ch === '}') {
      depth -= 1
      if (depth === 0 && start !== null) {
        objs.push(arr_text.slice(start + 1, i))
        start = null
      }
    }
  }
  return objs
}

function find_children_span(obj_text) {
  const m = /\bchildren\s*:\s*\[/.exec(obj_text)
  if (!m) return [null, obj_text]
  let depth = 1
  let i = m.index + m[0].length
  const mend = i
  while (i < obj_text.length && depth) {
    if (obj_text[i] === '[') depth += 1
    else if (obj_text[i] === ']') depth -= 1
    i += 1
  }
  return [obj_text.slice(mend, i - 1), obj_text.slice(0, m.index) + obj_text.slice(i)]
}

function parse_router_objects(arr_text, parent_path, chain, routes) {
  for (const obj of split_objects(arr_text)) {
    const [children_text, own] = find_children_span(obj)
    const a = attrs_of(own.replaceAll('path:', 'path=').replaceAll('element:', 'element={').replaceAll('index:', 'index '))
    // object syntax needs its own field regexes; redo path/index simply:
    const pm = /\bpath\s*:\s*["']([^"']+)["']/.exec(own)
    a.path = pm ? pm[1] : null
    a.index = Boolean(/\bindex\s*:\s*true/.exec(own))
    const em = /\belement\s*:\s*<\s*([A-Za-z_]\w*)/.exec(own)
    a.component = em ? em[1] : null
    const nm = ATTR_NAVIGATE.exec(own)
    a.navigate_to = nm ? ((nm[1] ?? nm[2]) ?? null) : null
    let full
    if (a.index) {
      full = parent_path || '/'
    } else if (a.path !== null) {
      full = join_paths(parent_path, a.path)
    } else {
      full = parent_path // pathless layout
    }
    if (children_text !== null) {
      const new_chain = chain.concat(a.component ? [a.component] : [])
      parse_router_objects(children_text, a.path !== null ? full : parent_path, new_chain, routes)
      continue // layout: not itself a route
    }
    if (a.path === null && !a.index) continue
    const entry = { path: full || '/', component: a.component, layout_chain: chain, audience: 'user' }
    if (a.navigate_to) {
      entry.audience = 'redirect'
      entry.redirect_to = a.navigate_to
    }
    if (a.path === '*') entry.not_found = true
    routes.push(entry)
  }
}

function find_entry(repo) {
  const cands = []
  const has = (f) => cands.includes(f)
  for (const name of ['src/App.tsx', 'src/App.jsx', 'src/main.tsx', 'src/main.jsx', 'src/router.tsx', 'src/routes.tsx', 'app/App.tsx']) {
    const f = path.join(repo, name)
    if (isFile(f)) cands.push(f)
  }
  for (const f of glob(repo, 'src/**/*.[tj]sx')) {
    if (!partsHasNodeModules(f) && !has(f)) {
      let t
      try { t = readTextReplace(f) } catch { continue }
      if (t.includes('createBrowserRouter') || t.includes('<Routes')) cands.push(f)
    }
  }
  for (const f of cands) {
    const t = readTextReplace(f)
    if (t.includes('createBrowserRouter') || t.includes('<Routes') || t.includes('<Route')) {
      return [f, t]
    }
  }
  return [null, null]
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------
function print(s) { fs.writeSync(1, s + '\n') }
function eprint(s) { fs.writeSync(2, s + '\n') }

// sys.exit("message"): print message to stderr, exit status 1.
function sysExitStr(msg) {
  eprint(msg)
  process.exit(1)
}

function usageLine(prog) {
  return `usage: ${prog} [-h] [--repo-root REPO_ROOT] [--entry ENTRY] [-o OUT]`
}
function argError(prog, msg) {
  eprint(usageLine(prog))
  eprint(`${prog}: error: ${msg}`)
  process.exit(2)
}
function parseArgs(argv, prog) {
  const args = { repo_root: '.', entry: null, out: 'map/_routes.json' }
  const positionals = []
  let i = 0
  const need = (name) => {
    if (i + 1 >= argv.length) argError(prog, `argument ${name}: expected one argument`)
    return argv[++i]
  }
  for (; i < argv.length; i++) {
    const a = argv[i]
    if (a === '-h' || a === '--help') { print(usageLine(prog)); process.exit(0) }
    else if (a === '--repo-root') args.repo_root = need('--repo-root')
    else if (a.startsWith('--repo-root=')) args.repo_root = a.slice('--repo-root='.length)
    else if (a === '--entry') args.entry = need('--entry')
    else if (a.startsWith('--entry=')) args.entry = a.slice('--entry='.length)
    else if (a === '-o' || a === '--out') args.out = need(a)
    else if (a.startsWith('--out=')) args.out = a.slice('--out='.length)
    else if (a.startsWith('-o') && a.length > 2) args.out = a.slice(2)
    else if (a.startsWith('-') && a !== '-') argError(prog, `unrecognized arguments: ${a}`)
    else positionals.push(a)
  }
  if (positionals.length > 0) argError(prog, `unrecognized arguments: ${positionals.join(' ')}`)
  return args
}

function main() {
  const prog = baseName(process.argv[1] || 'routes_react_router.mjs')
  const args = parseArgs(process.argv.slice(2), prog)
  const repo = path.resolve(args.repo_root)
  const aliases = load_aliases(repo)

  let entry, text
  if (args.entry) {
    entry = path.join(repo, args.entry)
    text = readTextReplace(entry)
  } else {
    ;[entry, text] = find_entry(repo)
    if (entry === null) {
      sysExitStr('No react-router registration found (createBrowserRouter or <Routes>). Wrong repo or wrong strategy.')
    }
  }

  const routes = []
  const m = /createBrowserRouter\s*\(\s*\[/.exec(text)
  if (m) {
    let depth = 1
    let i = m.index + m[0].length
    const mend = i
    while (i < text.length && depth) {
      if (text[i] === '[') depth += 1
      else if (text[i] === ']') depth -= 1
      i += 1
    }
    parse_router_objects(text.slice(mend, i - 1), '', [], routes)
  }
  routes.push(...parse_jsx(text))

  const imp = build_import_map(text)
  const out_routes = []
  const seen = new Set()
  for (const r of routes) {
    if (seen.has(r.path)) continue
    seen.add(r.path)
    const comp = 'component' in r ? r.component : null
    delete r.component
    let src_file = null
    if (comp && Object.prototype.hasOwnProperty.call(imp, comp)) {
      const resolved = resolve_import(entry, imp[comp], repo, aliases)
      if (resolved) {
        src_file = relTo(repo, resolved)
      }
    }
    const rec = {
      path: r.path,
      source_file: src_file || relTo(repo, entry),
      component: comp,
      layout_chain: 'layout_chain' in r ? r.layout_chain : [],
      auth_required: null, // react-router guards are wrapper components; decide per-app in Phase 1 prose
      audience: 'audience' in r ? r.audience : 'user',
    }
    if ('redirect_to' in r) rec.redirect_to = r.redirect_to
    if (r.not_found) rec.not_found = true
    rec.slug = route_slug(r.path)
    out_routes.push(rec)
  }
  const outp = args.out
  fs.mkdirSync(path.dirname(outp) || '.', { recursive: true })
  fs.writeFileSync(outp, dumps(out_routes, 2))
  const nred = out_routes.filter((r) => r.audience === 'redirect').length
  print(`react-router: ${out_routes.length} routes -> ${pyPosixStr(outp)} (${nred} redirects) from ${baseName(entry)}`)
  if (out_routes.length === 0) {
    sysExitStr('Router file found but zero routes parsed - inspect it by hand before continuing.')
  }
}

function isDirectRun() {
  try {
    return process.argv[1] && fs.realpathSync(process.argv[1]) === fs.realpathSync(fileURLToPath(import.meta.url))
  } catch {
    return import.meta.url === `file://${process.argv[1]}`
  }
}

if (isDirectRun()) {
  main()
}
